pkg/session/blocklist.go

```diff
@@ -0,0 +1,163 @@
+// everest
+// Copyright (C) 2025 Percona LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package session
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strconv"
+	"time"
+
+	"github.com/cenkalti/backoff/v4"
+	"github.com/golang-jwt/jwt/v5"
+	"go.uber.org/zap"
+	corev1 "k8s.io/api/core/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	dataKey         = "list"
+	maxRetries      = 10
+	backoffInterval = 500 * time.Millisecond
+)
+
+var (
+	errExtractJti       = errors.New("could not extract jti")
+	errExtractExp       = errors.New("could not extract exp")
+	errEmptyToken       = errors.New("token is empty")
+	errUnsupportedClaim = func(claims any) error {
+		return errors.New(fmt.Sprintf("unsupported claims type: %T", claims))
+	}
+)
+
+// Blocklist represents interface to block JWT tokens and check if a token is blocked.
+type Blocklist interface {
+	// Block invalidates the token from the context by adding it to blocklist.
+	Block(ctx context.Context, token *jwt.Token) error
+	// IsBlocked checks if the token from the context is blocked.
+	IsBlocked(ctx context.Context, token *jwt.Token) (bool, error)
+}
+
+type blocklist struct {
+	tokenStore TokenStore
+	l          *zap.SugaredLogger
+}
+
+// TokenStore represents an abstraction for storage, hiding details about how the data is actually stored.
+type TokenStore interface {
+	// Add adds the shortened token to the blocklist
+	Add(ctx context.Context, shortenedToken string) error
+	// Exists checks if the shortened token is in the blocklist
+	Exists(ctx context.Context, shortenedToken string) (bool, error)
+}
+
+// BlocklistClient supports only the k8s API methods that are needed for blocklist management.
+// A separate client is needed to apply the controller-runtime cache only to the related objects.
+// Using the controller-runtime client is also beneficial because it supports HA mode.
+type BlocklistClient interface {
```

Comment:
Secrets are part of TokenStore and are required in it's context only, blockList knows nothing about Secrets, it operates with TokenStore only.
it seems this is more `TokenStoreSecretClient` rathe then `BlocklistClient`

Reply:
The `BlocklistClient` defines the list of whatever external methods we need to make the current implementation of the Blocklist work. Otherwise we would assume that our global context allows us to make those external requests (that there is the k8s API available). So having the `BlocklistClient` is a dependency injection that allows the blocklist operate independently from the environment

Reply:
The list of the methods which are needed to make the Blocklist work may change with time. That's why I would prefer not specifying "Secrets" in the name

Reply:
This thread https://github.com/percona/everest/pull/1293#discussion_r2063029462 changed my mind, moved the dependencies inside the session manager and made it coupled with k8s API.